File: agent/plans/quality.py
from __future__ import annotations

from dataclasses import dataclass

from .models import PlanDraftState, PlanStep


@dataclass(frozen=True)
class PlanQualityResult:
    ok: bool
    errors: list[str]


class PlanQualityError(ValueError):
    def __init__(self, errors: list[str]) -> None:
        self.errors = list(errors)
        super().__init__(format_plan_quality_error(errors))


class PlanQualityGate:
    """Rejects weak plans before they become approval cards."""

    def assess(
        self,
        *,
        steps: list[PlanStep],
        draft: PlanDraftState,
    ) -> PlanQualityResult:
        errors: list[str] = []
        if not steps:
            errors.append("plan has no structured steps")
            return PlanQualityResult(ok=False, errors=errors)

        has_discoveries = bool(draft.discoveries)
        has_draft_verification = bool(draft.verification_strategy)
        for step in steps:
            errors.extend(_assess_step(
                step,
                has_discoveries=has_discoveries,
                has_draft_verification=has_draft_verification,
            ))
        return PlanQualityResult(ok=not errors, errors=errors)

    def require_ok(self, *, steps: list[PlanStep], draft: PlanDraftState) -> None:
        result = self.assess(steps=steps, draft=draft)
        if not result.ok:
            raise PlanQualityError(result.errors)


def format_plan_quality_error(errors: list[str]) -> str:
    lines = ["Error: plan quality gate failed"]
    lines.extend(f"- {error}" for error in errors)
    return "\n".join(lines)


def _assess_step(
    step: PlanStep,
    *,
    has_discoveries: bool,
    has_draft_verification: bool,
) -> list[str]:
    sid = step.id
    errors: list[str] = []
    if not _has_scope(step, has_discoveries=has_discoveries):
        errors.append(f"{sid} has no target files, discovery reference, or concrete scope")
    if _has_generic_title(step):
        errors.append(f"{sid} title is too generic; add concrete acceptance")
    if not _has_verification(step, has_draft_verification=has_draft_verification):
        errors.append(f"{sid} has no verification command or manual verification rule")
    if step.risk.strip().lower() == "high":
        if not step.risk_note.strip():
            errors.append(f"{sid} is high risk but has no risk note")
        if not step.rollback.strip():
            errors.append(f"{sid} is high risk but has no rollback path")
    return errors


def _has_scope(step: PlanStep, *, has_discoveries: bool) -> bool:
    if step.files or has_discoveries:
        return True
    if step.acceptance:
        return True
    return len(step.description.strip()) >= 24


def _has_generic_title(step: PlanStep) -> bool:
    title = " ".join(step.title.lower().strip().split())
    generic_titles = {
        "fix issue",
        "fix bug",
        "improve code",
        "update code",
        "make changes",
        "implement",
        "refactor",
        "修复问题",
        "优化代码",
        "改进代码",
    }
    return title in generic_titles and not step.acceptance


def _has_verification(step: PlanStep, *, has_draft_verification: bool) -> bool:
    return bool(step.commands or step.acceptance or has_draft_verification)

File: packages/core/src/plans/quality.ts
/**
 * PlanQualityGate (MIG-CTRL-013)。对齐 Python `agent/plans/quality.py`。
 * 在计划成为审批卡前拒绝弱计划：scope / 未知 discovery / 泛标题 / 验证 / 高风险 note+rollback。
 */
import type { PlanDraftState, PlanStep } from './models'

export interface PlanQualityResult {
  ok: boolean
  errors: string[]
}

export class PlanQualityError extends Error {
  readonly errors: string[]
  constructor(errors: string[]) {
    super(formatPlanQualityError(errors))
    this.name = 'PlanQualityError'
    this.errors = [...errors]
  }
}

export class PlanQualityGate {
  assess(opts: {
    steps: PlanStep[]
    draft: PlanDraftState
  }): PlanQualityResult {
    const errors: string[] = []
    if (!opts.steps.length) {
      errors.push('plan has no structured steps')
      return { ok: false, errors }
    }
    errors.push(...planTopologyErrors(opts.steps))
    const discoveryIds = new Set<string>()
    for (const item of opts.draft.discoveries) {
      if (item && typeof item === 'object') {
        const id = String((item as Record<string, unknown>).id ?? '').trim()
        if (id) discoveryIds.add(id)
      }
    }
    const hasDraftVerification = opts.draft.verificationStrategy.length > 0
    for (const step of opts.steps) {
      errors.push(...assessStep(step, discoveryIds, hasDraftVerification))
    }
    return { ok: errors.length === 0, errors }
  }

  requireOk(opts: { steps: PlanStep[]; draft: PlanDraftState }): void {
    const result = this.assess(opts)
    if (!result.ok) throw new PlanQualityError(result.errors)
  }
}

export function planTopologyErrors(
  steps: readonly PlanStep[],
  isSkippedDependencyWaived: (step: PlanStep) => boolean = () => false,
): string[] {
  const errors: string[] = []
  const byId = new Map<string, PlanStep>()
  for (const step of steps) {
    if (byId.has(step.id)) errors.push(`duplicate plan step id: ${step.id}`)
    else byId.set(step.id, step)
  }
  const active = steps.filter((step) => step.status === 'active')
  if (active.length > 1)
    errors.push('plan must contain at most one active step')
  for (const step of steps) {
    for (const dependencyId of step.dependsOn) {
      if (dependencyId === step.id) {
        errors.push(`${step.id} cannot depend on itself`)
        continue
      }
      if (!byId.has(dependencyId))
        errors.push(`${step.id} references unknown dependency: ${dependencyId}`)
    }
  }
  const visiting = new Set<string>()
  const visited = new Set<string>()
  const visit = (stepId: string): void => {
    if (visiting.has(stepId)) {
      errors.push(`plan dependency cycle includes ${stepId}`)
      return
    }
    if (visited.has(stepId)) return
    visiting.add(stepId)
    for (const dependencyId of byId.get(stepId)?.dependsOn ?? []) {
      if (byId.has(dependencyId)) visit(dependencyId)
    }
    visiting.delete(stepId)
    visited.add(stepId)
  }
  for (const step of steps) visit(step.id)
  for (const step of steps) {
    if (step.status !== 'active' && step.status !== 'done') continue
    for (const dependencyId of step.dependsOn) {
      const dependency = byId.get(dependencyId)
      if (
        dependency &&
        dependency.status !== 'done' &&
        !(
          dependency.status === 'skipped' &&
          isSkippedDependencyWaived(dependency)
        )
      ) {
        errors.push(
          `${step.id} dependency ${dependencyId} is not done or explicitly waived`,
        )
      }
    }
  }
  return [...new Set(errors)]
}

export function formatPlanQualityError(errors: string[]): string {
  return [
    'Error: plan quality gate failed',
    ...errors.map((e) => `- ${e}`),
  ].join('\n')
}

function assessStep(
  step: PlanStep,
  discoveryIds: Set<string>,
  hasDraftVerification: boolean,
): string[] {
  const sid = step.id
  const errors: string[] = []
  if (!hasScope(step, discoveryIds)) {
    errors.push(
      `${sid} has no target files, discovery reference, or concrete scope`,
    )
  }
  const unknownRefs = step.discoveryRefs.filter(
    (ref) => discoveryIds.size > 0 && !discoveryIds.has(ref),
  )
  if (unknownRefs.length) {
    errors.push(
      `${sid} references unknown discoveries: ${unknownRefs.slice(0, 3).join(', ')}`,
    )
  }
  if (hasGenericTitle(step)) {
    errors.push(`${sid} title is too generic; add concrete acceptance`)
  }
  if (!hasVerification(step, hasDraftVerification)) {
    errors.push(
      `${sid} has no verification command or manual verification rule`,
    )
  }
  if (step.risk.trim().toLowerCase() === 'high') {
    if (!step.riskNote.trim())
      errors.push(`${sid} is high risk but has no risk note`)
    if (!step.rollback.trim())
      errors.push(`${sid} is high risk but has no rollback path`)
  }
  return errors
}

function hasScope(step: PlanStep, discoveryIds: Set<string>): boolean {
  if (step.files.length) return true
  if (
    step.discoveryRefs.length &&
    (discoveryIds.size === 0 ||
      step.discoveryRefs.some((ref) => discoveryIds.has(ref)))
  ) {
    return true
  }
  if (step.acceptance.length) return true
  return step.description.trim().length >= 24
}

const GENERIC_TITLES = new Set([
  'fix issue',
  'fix bug',
  'improve code',
  'update code',
  'make changes',
  'implement',
  'refactor',
  '修复问题',
  '优化代码',
  '改进代码',
])

function hasGenericTitle(step: PlanStep): boolean {
  const title = step.title
    .toLowerCase()
    .trim()
    .split(/\s+/)
    .filter((p) => p)
    .join(' ')
  return GENERIC_TITLES.has(title) && !step.acceptance.length
}

function hasVerification(
  step: PlanStep,
  hasDraftVerification: boolean,
): boolean {
  return Boolean(
    step.commands.length || step.acceptance.length || hasDraftVerification,
  )
}
